pkg/mcp/sink/server_sink_test.go

```diff
@@ -37,29 +37,116 @@ type serverHarness struct {
-	return h, h.openError()
-}
-
 // avoid ambiguity between grpc.ServerStream and test.sinkTestHarness
 func (h *serverHarness) Context() context.Context {
 	return h.sinkTestHarness.Context()
 }
 
+func TestServerSinkRateLimitter(t *testing.T) {
+	ctx, _ := context.WithCancel(context.Background())
+	type key int
+	const ctxKey key = 0
+	expectedCtx := "expectedCtx"
+
+	testHarness := newSinkTestHarness()
+	testHarness.ctx = context.WithValue(ctx, ctxKey, expectedCtx)
+
+	h := &serverHarness{
+		sinkTestHarness: testHarness,
+	}
+
+	sinkOptions := &Options{
+		CollectionOptions: CollectionOptionsFromSlice(test.SupportedCollections),
+		Updater:           h,
+		ID:                test.NodeID,
+		Metadata:          test.NodeMetadata,
+		Reporter:          monitoring.NewInMemoryStatsContext(),
+	}
+
+	authChecker := test.NewFakeAuthChecker()
+	fakeLimiter := test.NewFakeRateLimiter()
+	close(fakeLimiter.WaitErr)
+
+	serverOpts := &ServerOptions{
+		AuthChecker: authChecker,
+		RateLimiter: fakeLimiter,
+	}
+
+	s := NewServer(sinkOptions, serverOpts)
+
+	go s.EstablishResourceStream(h)
+	c := <-fakeLimiter.WaitCh
+
+	ctxVal := c.Value(ctxKey).(string)
+	if ctxVal != expectedCtx {
+		t.Errorf("Wait received wrong context: got %v want %v", ctxVal, expectedCtx)
+	}
+
+	if len(fakeLimiter.WaitCh) != 0 {
+		t.Error("Stream() => expected Wait to only get called once")
+	}
+}
+
+func TestServerSinkRateLimitterError(t *testing.T) {
+	testHarness := newSinkTestHarness()
+
+	h := &serverHarness{
+		sinkTestHarness: testHarness,
+	}
+
+	sinkOptions := &Options{
+		CollectionOptions: CollectionOptionsFromSlice(test.SupportedCollections),
+		Updater:           h,
+		ID:                test.NodeID,
+		Metadata:          test.NodeMetadata,
+		Reporter:          monitoring.NewInMemoryStatsContext(),
+	}
+
+	authChecker := test.NewFakeAuthChecker()
+	fakeLimiter := test.NewFakeRateLimiter()
+
+	serverOpts := &ServerOptions{
+		AuthChecker: authChecker,
+		RateLimiter: fakeLimiter,
+	}
+
+	s := NewServer(sinkOptions, serverOpts)
+
+	expectedErr := "rate limiting gone wrong"
+	fakeLimiter.WaitErr <- errors.New("rate limiting gone wrong")
+
+	errC := make(chan error)
+	go func() {
+		errC <- s.EstablishResourceStream(h)
+	}()
+
+	err := <-errC
+	if err == nil || err.Error() != expectedErr {
+		t.Fatalf("Expected error from Wait: got %v want %v ", err, expectedErr)
+	}
+}
 func TestServerSink(t *testing.T) {
 	h := &serverHarness{
 		sinkTestHarness: newSinkTestHarness(),
 	}
 
 	authChecker := test.NewFakeAuthChecker()
-	options := &Options{
+	sinkOptions := &Options{
 		CollectionOptions: CollectionOptionsFromSlice(test.SupportedCollections),
 		Updater:           h,
 		ID:                test.NodeID,
 		Metadata:          test.NodeMetadata,
 		Reporter:          monitoring.NewInMemoryStatsContext(),
 	}
-	s := NewServer(options, &ServerOptions{AuthChecker: authChecker})
+
+	rateLimiter := test.NewFakeRateLimiter()
+	close(rateLimiter.WaitErr)
```

Comment:
Maybe a comment that this forces the rate limiter to always return `nil` since reads from closed channels return the `nil` value?